src/data_structures/bwt.rs

```diff
@@ -123,11 +123,46 @@ impl Occ {
     /// Get occurrence count of symbol a in BWT[..r+1].
     /// Complexity: O(k).
     pub fn get(&self, bwt: &BWTSlice, r: usize, a: u8) -> usize {
-        let i = r / self.k;
-
-        let count = bwt[(i * self.k) + 1..r + 1].iter().filter(|&&c| c == a).count();
+        // NOTE:
+        //
+        // Retrieving byte match counts in this function is critical to the performance of FM Index.
+        //
+        // The below manual count code is roughly equivalent to:
+        // ```
+        // let count = bwt[(i * self.k) + 1..r + 1].iter().filter(|&&c| c == a).count();
+        // self.occ[i][a as usize] + count
+        // ```
+        //
+        // But there are a couple of reasons to do this manually:
+        // 1) As of 2016, versions of rustc/LLVM vectorize this manual loop more reliably
+        //    than the iterator adapter version.
+        // 2) Manually accumulating the byte match count in a single chunk can allows
+        //    us to use a `u32` for that count, which has faster arithmetic on common arches.
+        //    This does necessitate storing `k` as a u32.
+        //
+        // See the conversation in these issues for some of the history here:
+        //
+        // https://github.com/rust-bio/rust-bio/pull/74
+        // https://github.com/rust-bio/rust-bio/pull/76
+
+        // self.k is our sampling rate, so find our last sampled checkpoint
+        let i = r / self.k as usize;
+        let checkpoint = self.occ[i][a as usize];
+
+        // find the portion of the BWT past the checkpoint which we need to count
+        let start = (i * self.k as usize) + 1;
+        let end = r + 1;
+
+        // count all the matching bytes b/t the closest checkpoint and our desired lookup
+        let mut count: u32 = 0;
+        for &x in &bwt[start..end] {
+            if x == a {
+                count += 1;
```

Comment:
What happens if I do `count += (x == a) as u32`? I think it should be cheaper, because addition is faster than branch prediction?